### PR TITLE
Support record compression

### PR DESCRIPTION
**Pull request by CrazyHackGUT**
- Implemented "out-of-box" 3 compression methods: BZip, GZip and Zip
- Refactoring: moved methods isAdmin() and loggedUser() from \App\Controller\AbstractController to \App
- Fixed bug when server id `0` can cause error `Invalid key` from API when key is valid

### CODE REVIEW

**Comment by CrazyHackGUT** [comment]
Need extra attention for GitHub Actions: Psalm don't see optional ZIP extension.

File: src/App.php [hunk 1]
```diff
@@ -491,4 +509,23 @@ public function isInstalled(): bool
 
         return $success;
     }
+
+    public function isAdmin(): bool
+    {
+        $administrators = $this->config()['system']['administrators'] ?? [];
```
Comment (West14) [inline src/App.php:515]: Maybe will be better if we add the `administrators` field to the default config, In this case, we could remove null check.
What do you think?

File: src/App/Util/Demo.php [hunk 2]
```diff
@@ -32,26 +47,50 @@ public static function cleanup(App $app): array
...
         }
 
         if (!empty($demoIds))
         {
             $escapedDemoIds = array_map(function ($el) { return "'$el'"; }, $demoIds);
-            $db->query(sprintf('DELETE FROM `record` WHERE `demo_id` IN (%s)', implode(', ', $escapedDemoIds)));
+            $escapedDemoIdsWithSeparator = implode(', ', $escapedDemoIds);
+
+            $db->query(sprintf('DELETE FROM `record` WHERE `demo_id` IN (%s)', $escapedDemoIdsWithSeparator));
+            $db->query(sprintf('DELETE FROM `record` WHERE `demo_id` IN (%s)', $escapedDemoIdsWithSeparator));
```
Comment (West14) [inline src/App/Util/Demo.php:71]: Just mistake marker.

**Review by West14 — CHANGES_REQUESTED** [review]

File: src/App.php (hunk 1 above)
Comment (CrazyHackGUT) [reply src/App.php:515]: `array_merge` just replaces values for all keys, without type checking and recursively replacing into sub-arrays, so this is currently not possible.
We need write own `array_merge`.

**Review by West14 — APPROVED** [review]

**Review by West14 — APPROVED** [review]

**Review by West14 — APPROVED** [review]

**Pull request merged**